### ISSUE TITLE
[FeatureRequest] Support a `__post_init__` hook similar to `dataclasses`

### ISSUE THREAD

**Issue by dhirschfeld** [jcrist/msgspec#70]
***xref:*** https://docs.python.org/3/library/dataclasses.html#post-init-processing

**Comment by dhirschfeld** [comment]
Sorry, meant to provide more info but got distracted! Will come back to this to give more background...

**Comment by dhirschfeld** [comment]
I'm just trying to learn about `msgspec` so what I'm trying to do might be silly. In any case, I was trying to see if you could create a class that knew how to serialize and deserialize itself. My first attempt was:
```python
class SerializationBase(object):
    __decoder: msgspec.msgpack.Decoder
    __encoder = msgspec.msgpack.Encoder()
    
    def __init__(self):
        self.__decoder = msgspec.msgpack.Decoder(type(self))

    def dumps(self):
        return self.__encoder.encode(self)

    def loads(self, data: bytes) -> Any:
        return self.__decoder.decode(data)


class Param(SerializationBase, msgspec.Struct):
    param: int


class ParamList(SerializationBase, msgspec.Struct):
    params: List[Param]
```
...but that resulted in:
```
TypeError: All base classes must be subclasses of msgspec.Struct
```
So, `SerializationBase` needs to inherit from ` msgspec.Struct` which means I can't initialise the decoder in the `__init__` with `type(self)` 🤔 

If I defined the decoder on-the-fly then this actually worked!
```python
class SerializationBase(msgspec.Struct):
    __encoder = msgspec.msgpack.Encoder()

    def dumps(self):
        return self.__encoder.encode(self)

    def loads(self, data: bytes) -> Any:
        return msgspec.msgpack.decode(data, type=type(self))
```
```python
>>> params = ParamList([Param(param=n) for n in range(3)])
>>> params
ParamList(params=[Param(param=0), Param(param=1), Param(param=2)])

>>> data = params.dumps()
>>> data
b'\x81\xa6params\x93\x81\xa5param\x00\x81\xa5param\x01\x81\xa5param\x02'

>>> params.loads(data)
ParamList(params=[Param(param=0), Param(param=1), Param(param=2)])
```

Since defining the typed decoder on-the-fly is likely to be inefficient I was trying to find a way to just create it once at class creation or initialisation and thought a `__post_init__` hook could do the job.

**Comment by dhirschfeld** [comment]
I also though I might be able to use `__init_subclasses__` but that just segfaulted!
```python
class SerializationBase(msgspec.Struct):
    __encoder = msgspec.msgpack.Encoder()

    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)
        cls.__decoder = msgspec.msgpack.Decoder(cls)

    def dumps(self):
        return self.__encoder.encode(self)

    def loads(self, data: bytes) -> Any:
        return self.__decoder.decode(data)

class Param(SerializationBase, msgspec.Struct):
    param: int
```

**Comment by dhirschfeld** [comment]
So, it at least can be done but what I'm really after is a `loads` function which can take data serialised by *any* class and return the corresponding class (without having to manually specify what class the data represents).

Does that make sense? Is it possible? Is it sensible?!

**Comment by jcrist** [comment]
Thanks for the example. The segfault there is due to the struct type itself not being fully initialized when `__init_subclass__` is called. Unfortunately due to the way cpython initializes types this isn't an easy fix (I've opened #73 to track it). It's doable, but will require moving a bunch of logic around.

For the pattern you're trying to do though, `__init_subclass__` should work - provided you don't have any recursive types (types that have fields containing themselves). Recursive types won't work, since the definition needs to be available when the decoder is created, but won't be yet since `__init_subclass__` hasn't finished running.

A better pattern would be to use a cached descriptor on the class itself to initialize the decoder on first use. This works now, and will also work with recursive types.

```python
import msgspec
from typing import Any


class delayed_decoder_attr:
    def __get__(self, _, cls=None):
        # This only runs once for each subclass,
        # overwriting itself on the cls.__dict__
        val = msgspec.msgpack.Decoder(cls)
        setattr(cls, "_decoder", val)
        return val


class Base(msgspec.Struct):
    _encoder = msgspec.msgpack.Encoder()

    def __init_subclass__(cls, **kwargs):
        # Create a new instance in each subclass
        # so that subclasses don't overwrite
        cls._decoder = delayed_decoder_attr()

    def dumps(self):
        return self._encoder.encode(self)

    @classmethod
    def loads(cls, data: bytes) -> Any:
        return cls._decoder.decode(data)


class Point(Base):
    x: int
    y: int


# Subclasses of subclasses also work
class Point3D(Point):
    x: int
    y: int
    z: int


print(Point.loads(Point(1, 2).dumps()))
print(Point3D.loads(Point3D(1, 2, 3).dumps()))
```

> but what I'm really after is a loads function which can take data serialised by any class and return the corresponding class (without having to manually specify what class the data represents).

This isn't really possible (at least currently). Protocols like JSON or MessagePack are less type-rich than something like python (for example, JSON only has one array type, while python has many sequence types you might want to map to that array type). `msgspec` uses python type annotations to describe how the mapping between JSON types and Python types should be done, since that mapping isn't part of the serialized message.

What you're after is a way to support decoding into a union of different struct types, with some way of automatically detecting what type to use. This isn't currently supported, but is something I'd like to support. The trick is that there are many different ways people serialize information required to disambiguate JSON objects. The serde docs have a nice example of common schemes: https://serde.rs/enum-representations.html. My plan for this is supporting the "internally tagged" and "adjacently tagged" schemes, both of which are common in JSON apis. I do not plan to support the "untagged" scheme, as it cannot be done in an efficient manner.

**Comment by dhirschfeld** [comment]
Thanks for the info @jcrist. That's a neat trick with the descriptor! I was playing around with descriptors and got something to work with a `decoder_registry: Dict[Type, msgspec.msgpack.Decoder]` which was used from the `__get__`. It seems more elegant to store the encoder on the class itself than in a separate registry though.


I'm happy with `__post_init__` not being implemented - I was more just trying to find a way to allow a class to decode itself. Since that's answered here, I'll go ahead and close...

**Comment by dhirschfeld** [comment]
***xref:*** Fully resolved in #470
